[FLINK-20354] Rework standalone docs pages (#14346)
apache/flink: docs/deployment/resource-providers/standalone/docker.md

under the License.
-->

* This will be replaced by the TOC
{:toc}

## Getting Started

This *Getting Started* section guides you through the local setup (on one machine, but in separate containers) of a Flink cluster using Docker containers.

### Introduction

[Docker](https://www.docker.com) is a popular container runtime.
There are Docker images for Apache Flink available [on Docker Hub](https://hub.docker.com/_/flink).

> You can use the Docker images to deploy a *Session* or *Application cluster* on Docker. This page focuses on the setup of Flink on Docker, Docker Swarm and Docker compose.

Reviewer: ```suggestion
You can use the Docker images to deploy a *Session* or *Application cluster* on Docker. This page focuses on the setup of Flink on Docker, Docker Swarm and Docker Compose.
```

    --network flink-network \
    --env FLINK_PROPERTIES="${FLINK_PROPERTIES}" \
    flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest{% endif %} taskmanager
```

The web interface is now available at [localhost:8081](http://localhost:8081).

Submission of a job is now possible like this (assuming you have a local distribution of Flink available):

```sh
$ ./bin/flink run ./examples/streaming/TopSpeedWindowing.jar
```

To shut down the cluster, either terminate (e.g. with `CTRL-C`) the JobManager and TaskManager processes, or use `docker ps` to identify and `docker stop` to terminate the containers.

## Deployment Modes

The Flink image contains a regular Flink distribution with its default configuration and a standard entry point script.
You can run its entry point in the following modes:
* [JobManager]({% link concepts/glossary.md %}#flink-jobmanager) for [a Session cluster](#starting-a-session-cluster-on-docker)
* [JobManager]({% link concepts/glossary.md %}#flink-jobmanager) for [a Application cluster](#application-mode-on-docker)
* [TaskManager]({% link concepts/glossary.md %}#flink-taskmanager) for any cluster

> This allows you to deploy a standalone cluster (Session or Per-Job Mode) in any containerised environment, for example:

Reviewer: ```suggestion
This allows you to deploy a standalone cluster (Session or Application Mode) in any containerised environment, for example:
```

* manually in a local Docker setup,

> * [in a (native) Kubernetes cluster]({% link deployment/resource-providers/standalone/kubernetes.md %}),

Reviewer: ```suggestion
* [in a Kubernetes cluster]({% link deployment/resource-providers/standalone/kubernetes.md %}),
```
I'd suggest removing `(native)` here as we're referring to the Standalone Kubernetes deployment. ...just to avoid confusion

    flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest{% endif %} taskmanager
```

The web interface is now available at [localhost:8081](http://localhost:8081).

Submission of a job is now possible like this (assuming you have a local distribution of Flink available)

```sh
./bin/flink run ./examples/streaming/TopSpeedWindowing.jar
```

To shut down the cluster, either terminate (e.g. with CTRL-C) the JobManager and TaskManager processes, or use `docker ps` to identify and `docker stop` to terminate the containers.

> ## Deployment Modes Supported on Docker

Author: I think approach 1 makes more sense.

The Flink image contains a regular Flink distribution with its default configuration and a standard entry point script.
You can run its entry point in the following modes:
* [JobManager]({% link concepts/glossary.md %}#flink-jobmanager) for [a Session cluster](#starting-a-session-cluster-on-docker)
* [JobManager]({% link concepts/glossary.md %}#flink-jobmanager) for [a Application cluster](#application-mode-on-docker)

> * [TaskManager]({% link concepts/glossary.md %}#flink-taskmanager) for any cluster

Author: I think it doesn't hurt, and it's more work removing the references than keeping them, so 🤷‍♂️

For example, you can use the following aliases:

* `flink:latest` → `flink:<latest-flink>-scala_<latest-scala>`
* `flink:1.11` → `flink:1.11.<latest-flink-1.11>-scala_2.11`

<span class="label label-info">Note</span> It is recommended to always use an explicit version tag of the docker image that specifies both the needed Flink and Scala
versions (for example `flink:1.11-scala_2.12`).
This will avoid some class conflicts that can occur if the Flink and/or Scala versions used in the application are different
from the versions provided by the docker image.

<span class="label label-info">Note</span> Prior to Flink 1.5 version, Hadoop dependencies were always bundled with Flink.
You can see that certain tags include the version of Hadoop, e.g. (e.g. `-hadoop28`).
Beginning with Flink 1.5, image tags that omit the Hadoop version correspond to Hadoop-free releases of Flink
that do not include a bundled Hadoop distribution.

### Passing configuration via environment variables

When you run Flink image, you can also change its configuration options by setting the environment variable `FLINK_PROPERTIES`:

> 

Author: I'm not sure if that really helps avoiding confusion. I would prefer to keep it this way.

If the main function of the user job main class accepts arguments, you can also pass them at the end of the `docker run` command.

### Session Mode on Docker

Local deployment in the session mode has already been described in the [introduction](#starting-a-session-cluster-on-docker) above.

### Flink with Docker Compose

[Docker Compose](https://docs.docker.com/compose/) is a way to run a group of Docker containers locally.
The next sections show examples of configuration files to run Flink.

#### Usage

* Create the `yaml` files with the container configuration, check examples for:
  * [Session cluster](#session-cluster-with-docker-compose)
  * [Application cluster](#application-cluster-with-docker-compose)

  See also [the Flink Docker image tags](#image-tags) and [how to customize the Flink Docker image](#advanced-customization)
  for usage in the configuration files.

> * Launch a cluster in the foreground

Author: I will shorten it a tiny bit

* **either mount a volume** with the custom configuration files to this path `/opt/flink/conf` when you run the Flink image:

    ```sh
    $ docker run \
        --mount type=bind,src=/host/path/to/custom/conf,target=/opt/flink/conf \
        flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest{% endif %} <jobmanager|standalone-job|taskmanager>
    ```

* or add them to your **custom Flink image**, build and run it:

>   *Dockerfile*:

Reviewer: ```suggestion
```
I think we can remove `Dockerfile:` here as it's mentioned in the item above that it's a "custom Flink image"

    docker exec -t -i "${JM_CONTAINER}" flink run -d -c ${JOB_CLASS_NAME} /job.jar
    ```

> #### Session Cluster with Docker Compose

Reviewer: Can't we just have a code block for each of the yaml configurations? Not sure whether the section are necessary here.

Reviewer: FYI: What about using collapsible blocks to hide long configuration files:
```
<p>
  <a class="btn" data-toggle="collapse" href="#session-cluster-yml" role="button" aria-expanded="false" aria-controls="session-cluster-yml">
    docker-compose.yml
  </a>
</p>
<div class="collapse" id="session-cluster-yml">
  {% highlight yaml %}
version: "2.2"
services:
  jobmanager:
    image: flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest{% endif %}
    ports:
      - "8081:8081"
    command: jobmanager
    environment:
      - |
        FLINK_PROPERTIES=
        jobmanager.rpc.address: jobmanager

  taskmanager:
    image: flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest{% endif %}
    depends_on:
      - jobmanager
    command: taskmanager
    scale: 1
    environment:
      - |
        FLINK_PROPERTIES=
        jobmanager.rpc.address: jobmanager
        taskmanager.numberOfTaskSlots: 2
  {% endhighlight %}
</div>
```
I got inspired by [this blog post](https://www.iditect.com/how-to/52576626.html).
```

Author: I'll try it out!

> 

Author: I think the text states already "jar file names", and the example is also without a path. In my opinion, this is sufficient.

There are also more [advanced ways](#advanced-customization) for customizing the Flink image.

### Enabling Python

> To build a custom image which has Python and Pyflink prepared, you can refer to the following Dockerfile:

Reviewer: ```suggestion
To build a custom image which has Python and PyFlink prepared, you can refer to the following Dockerfile:
```

* add other libraries to `/opt/flink/lib` (e.g. Hadoop)
* add other plugins to `/opt/flink/plugins`

You can customize the Flink image in several ways:

* **override the container entry point** with a custom script where you can run any bootstrap actions.
  At the end you can call the standard `/docker-entrypoint.sh` script of the Flink image with the same arguments
  as described in [supported deployment modes](#deployment-modes).

  The following example creates a custom entry point script which enables more libraries and plugins.

>   The custom script, custom library and plugin are provided from a mounted volume.

Reviewer: The linewraps look weird with the comments being on the same line as the commands in the code below. Can't we move the comments in front of the command on a separate line to avoid having linewraps?

    # create custom_lib.jar
    # create custom_plugin.jar

    $ echo "
    # enable an optional library
    ln -fs /opt/flink/opt/flink-queryable-state-runtime-*.jar /opt/flink/lib/
    # enable a custom library
    ln -fs /mnt/custom_lib.jar /opt/flink/lib/

    mkdir -p /opt/flink/plugins/flink-s3-fs-hadoop
    # enable an optional plugin
    ln -fs /opt/flink/opt/flink-s3-fs-hadoop-*.jar /opt/flink/plugins/flink-s3-fs-hadoop/  

    mkdir -p /opt/flink/plugins/custom_plugin
    # enable a custom plugin
    ln -fs /mnt/custom_plugin.jar /opt/flink/plugins/custom_plugin/

    /docker-entrypoint.sh <jobmanager|standalone-job|taskmanager>
    " > custom_entry_point_script.sh

    $ chmod 755 custom_entry_point_script.sh

    $ docker run \
        --mount type=bind,src=$(pwd),target=/mnt
        flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest{% endif %} /mnt/custom_entry_point_script.sh
    ```

* **extend the Flink image** by writing a custom `Dockerfile` and build a custom image:

>   *Dockerfile*:

Reviewer: ```suggestion
```
I think that's obsolete again.

    ENV VAR_NAME value
    ```

  **Commands for building**:

    ```sh
    docker build -t custom_flink_image .
    # optional push to your docker image registry if you have it,
    # e.g. to distribute the custom image to your cluster
    docker push custom_flink_image
    ```

### Flink with Docker Compose

[Docker Compose](https://docs.docker.com/compose/) is a way to run a group of Docker containers locally.
The next sections show examples of configuration files to run Flink.

#### Usage

* Create the `yaml` files with the container configuration, check examples for:
  * [Application cluster](#app-cluster-yml)

>   * [Session cluster](#session-cluster-yml)

Reviewer: I couldn't find the corresponding anchor points/yaml files. The targets of these links do not exist.

Reviewer: I just realized that you used the collapsible code blocks. Reading through it right now, I guess, I prefer the Appendix approach we use in the Kubernetes documentation page.

Author: Okay, I'll revert the collapse-thing.
I don't think adding another indirection to a appending makes sense here, since this is already the bottom of the page.

    ```sh
    $ docker-compose scale taskmanager=<N>
    ```

* Access the JobManager container

    ```sh
    $ docker exec -it $(docker ps --filter name=jobmanager --format={% raw %}{{.ID}}{% endraw %}) /bin/sh
    ```

* Kill the cluster

    ```sh
    $ docker-compose kill
    ```

* Access Web UI

  When the cluster is running, you can visit the web UI at [http://localhost:8081](http://localhost:8081).
  You can also use the web UI to submit a job to a *Session cluster*.

* To submit a job to a *Session cluster* via the command line, you can either

  * use [Flink CLI]({% link deployment/cli.md %}) on the host if it is installed:

    ```sh
    $ ./bin/flink run --detached --class ${JOB_CLASS_NAME} /job.jar
    ```

  * or copy the JAR to the JobManager container and submit the job using the [CLI]({% link deployment/cli.md %}) from there, for example:

    ```sh
    $ JOB_CLASS_NAME="com.job.ClassName"
    $ JM_CONTAINER=$(docker ps --filter name=jobmanager --format={% raw %}{{.ID}}{% endraw %}))
    $ docker cp path/to/jar "${JM_CONTAINER}":/job.jar
    $ docker exec -t -i "${JM_CONTAINER}" flink run -d -c ${JOB_CLASS_NAME} /job.jar
    ```

> Here, we provide the **docker-compose.yml:** <a id="app-cluster-yml">file</a> for *Application Cluster*.

Reviewer: ```suggestion
Here, we provide the <a id="app-cluster-yml">docker-compose.yml</a> for *Application Cluster*.
```

> Note: For the application cluster, the artifacts must be available in the Flink containers, check details [here](#application-mode-on-docker).

Reviewer: ```suggestion
Note: For the Application Mode cluster, the artifacts must be available in the Flink containers, check details [here](#application-mode-on-docker).
```

The `flink-conf.yaml` file must have write permission so that the Docker entry point script can modify it in certain cases.

### Using filesystem plugins

As described in the [plugins]({% link deployment/filesystems/plugins.md %}) documentation page: in order to use plugins they must be
copied to the correct location in the Flink installation in the Docker container for them to work.

If you want to enable plugins provided with Flink (in the `opt/` directory of the Flink distribution), you can pass the environment variable `ENABLE_BUILT_IN_PLUGINS` when you run the Flink image.
The `ENABLE_BUILT_IN_PLUGINS` should contain a list of plugin jar file names separated by `;`. A valid plugin name is for example `flink-s3-fs-hadoop-{{site.version}}.jar`

```sh
    docker run \
        --env ENABLE_BUILT_IN_PLUGINS=flink-plugin1.jar;flink-plugin2.jar \
        flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest{% endif %} <jobmanager|standalone-job|taskmanager>
```

There are also more [advanced ways](#advanced-customization) for customizing the Flink image.

> ### Enabling Python

Reviewer: Should we point to the Python CLI that is going to get introduced by @shuiqiangchen in [this PR](https://github.com/XComp/flink/pull/1)?

Reviewer 2: Maybe we should add a link to guide users to this part for kubernetes application submission which requires users to provide an image in the Python CLI section.

Author: Will this PR also get opened against Flink?

Reviewer: Yup, here's the [PR #14367 against Flink's master](https://github.com/apache/flink/pull/14367).

    taskmanager
```

The *job artifacts* must be available in the JobManager container, as outlined [here](#application-mode-on-docker).
See also [how to specify the JobManager arguments](#jobmanager-additional-command-line-arguments) to pass them
to the `flink-jobmanager` container.

The example assumes that you run the swarm locally and expects the *job artifacts* to be in `/host/path/to/job/artifacts`.

> It also mounts the host path with the artifacts as a volume to the container's path `/opt/flink/usrlib`.    

Reviewer: ```suggestion
It also mounts the host path with the artifacts as a volume to the container's path `/opt/flink/usrlib`.
```

Author: Sorry for this stuff. I should check the diff myself.
